File: 6.12/main.c
#include "../nob.h"

typedef struct {
	string *items;
	size_t count;
	size_t capacity;
} LINES;

static string trim_copy(const string s) {
    const string start = s;
    const string end = s + strlen(s);

    const char *b = start;
    while (*b && isspace((uch)*b)) b++;

    const char *e = end;
    while (e > b && isspace((uch)*(e - 1))) e--;

    size_t len = (size_t)(e - b);
    string out = (string)malloc(len + 1);
    memcpy(out, b, len);
    out[len] = '\0';
    return out;
}

static ll parse_ll(const string s) {
    ll v = 0;
     const char *p = s;
    while (*p) {
        if (isdigit((uch)*p)) v = v * 10 + (*p - '0');
        p++;
    }
    return v;
}

int main(int argc, char **argv) {

	FILE *input = open_input_or_die(argc, argv);

    size_t rows = 0, cols = 0;
    
	LINES lines = {0};

	ll part1 = 0, part2 = 0;

	string line = NULL;
	size_t size = 0;
	ssize_t read = 0;

	while((read = getline_or_die(&line, &size, input)) != -1) {
		size_t len = strlen(line);

		if(len && line[len-1] == '\n') line[--len] = '\0';

		string copy = (string)malloc(len + 1);
		strcpy(copy, line);

	    if (len > cols) cols = len;

		da_append(&lines, copy);
				
		rows++;	
	}

    for (size_t i = 0; i < rows; i++) {
        size_t cur = strlen(lines.items[i]);
        if (cur < cols) {
            lines.items[i] = (string)realloc(lines.items[i], cols + 1);
            memset(lines.items[i] + cur, ' ', cols - cur);
            lines.items[i][cols] = '\0';
        }
    }

	size_t op_row = rows - 1;
	int *col_has = (int *)calloc(cols, sizeof(int));

	for (size_t c = 0; c < cols; c++){
	    for (size_t r = 0; r < rows; r++){
	        if (lines.items[r][c] != ' ') col_has[c] = 1;
        }
    }

	for (size_t c = 0; c < cols;) {
	    while (c < cols && !col_has[c]) c++;
	    if (c >= cols) break;

	    size_t l = c;
	    while (c < cols && col_has[c]) c++;
	    size_t r = c - 1;

	    char op = '+';
	    for (size_t cc = l; cc <= r; cc++) {
	        if (lines.items[op_row][cc] != ' ') op = lines.items[op_row][cc];
        }

	    // PART 1 (TOP → BOTTOM, ROW BASED)
	    ll acc1 = (op == '*') ? 1 : 0;
	    int used1 = 0;

	    for (size_t row = 0; row < op_row; row++) {
	        size_t len = r - l + 1;
	        string sub = (string)malloc(len + 1);
	        memcpy(sub, lines.items[row] + l, len);
	        sub[len] = '\0';

	        string trim = trim_copy(sub);
	        free(sub);

	        if (*trim) {
	            ll v = parse_ll(trim);
	            used1 = 1;
	            if (op == '+') acc1 += v;
	            else acc1 *= v;
	        }

	        free(trim);
	    }

	    if (used1) part1 += acc1;

	    // PART 2 (RIGHT → LEFT, COLUMN BASED)
	    ll acc2 = (op == '*') ? 1 : 0;
	    int used2 = 0;
    	
	    for (size_t cc = r + 1; cc-- > l;) {
			DA buf = {0};

	        for (size_t row = 0; row < op_row; row++) {
	            if (lines.items[row][cc] != ' ') da_append(&buf,lines.items[row][cc]);
            }

			da_append(&buf, '\0');

			if (buf.count) {
	        	ll v = parse_ll(buf.items);
	            used2 = 1;
	            if (op == '+') acc2 += v;
	            else acc2 *= v;
	        }

	        da_free(&buf);
	    }

	    if (used2) part2 += acc2;
	}

    printf("Part 1: %llu\n", (ull) part1);
    printf("Part 2: %llu\n", (ull) part2);

	da_free_deep(&lines);
    free(line);
    free(col_has);
    fclose(input);

    return 0;
}

File: 6.12/main.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead, BufReader};
use std::str;

// Parse a string containing digits into i64
fn parse_ll(s: &str) -> i64 {
    let mut v: i64 = 0;
    for b in s.bytes() {
        if b.is_ascii_digit() {
            v = v * 10 + (b - b'0') as i64;
        }
    }
    v
}

fn main() -> io::Result<()> {

    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <input_file>", args[0]);
        std::process::exit(1);
    }

    let file = File::open(&args[1])?;
    let reader = BufReader::new(file);

    let mut lines: Vec<String> = Vec::new();
    let mut cols = 0;

    for line_res in reader.lines() {

        let line = line_res?;

        if line.len() > cols {
            cols = line.len();
        }

        lines.push(line);
    }

    let rows = lines.len();

    for line in &mut lines {
        while line.len() < cols {
            line.push(' ');
        }
    }

    let op_row = rows - 1;
    let mut col_has = vec![false; cols];

    for c in 0..cols {
        for r in 0..op_row {

            if lines[r].as_bytes()[c] != b' ' {
                col_has[c] = true;
                break;
            }
        }
    }

    let mut part1: i64 = 0;
    let mut part2: i64 = 0;

    let mut c = 0;
    while c < cols {

        while c < cols && !col_has[c] {
            c += 1;
        }

        if c >= cols {
            break;
        }

        let l = c;

        while c < cols && col_has[c] {
            c += 1;
        }

        let r = c - 1;

        let mut op = b'+';

        for cc in l..=r {

            let ch = lines[op_row].as_bytes()[cc];

            if ch != b' ' {
                op = ch;
            }
        }

        // -------- Part 1: Top-to-bottom column numbers --------
        let mut acc1 = if op == b'*' { 1 } else { 0 };
        let mut used1 = false;

        for row in 0..op_row {

            let len = r - l + 1;
            let sub = &lines[row][l..l + len];
            let trim = sub.trim();

            if !trim.is_empty() {
                let v = parse_ll(trim);
                used1 = true;

                if op == b'+' {
                    acc1 += v;
                } else {
                    acc1 *= v;
                }
            }
        }

        if used1 {
            part1 += acc1;
        }

        // -------- Part 2: Right-to-left column numbers --------
        let mut acc2 = if op == b'*' { 1 } else { 0 };
        let mut used2 = false;

        for cc in (l..=r).rev() {
            let mut buf: Vec<u8> = Vec::new();

            for row in 0..op_row {
                let ch = lines[row].as_bytes()[cc];

                if ch != b' ' {
                    buf.push(ch);
                }
            }
            if !buf.is_empty() {
                let s = str::from_utf8(&buf).unwrap();
                let v = parse_ll(s);
                used2 = true;

                if op == b'+' {
                    acc2 += v;
                } else {
                    acc2 *= v;
                }
            }
        }
        
        if used2 {
            part2 += acc2;
        }
    }

    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);

    Ok(())
}
